proposals/NNNN-noncopyable-generics.md

```diff
@@ -0,0 +1,713 @@
+struct S: CopyableProtocol, ~Copyable { ... }
+```
+
+This diagnostic only occurs if the `~Copyable` is directly specified:
+```swift
+protocol CopyableProtocol /* : Copyable */ { ... }
+protocol MaybeCopyableProtocol: ~Copyable { ... }
+struct S: CopyableProtocol, MaybeCopyableProtocol { ... } // OK
+```
+
+### Scoping rule
+
+A constraint suppression like `~Copyable` can only be applied to a type parameter within the same scope as the constraint:
+
+```swift
+struct S<T> { // signature: <T: Copyable>
+  func f() where T: ~Copyable // signature: <T>
+  // error: cannot suppress constraint 'T: ~Copyable' on generic parameter 'T' defined in outer scope
+}
+```
+
+Rationale: An outer generic context, like `S<T>`, already requires that `T` is `Copyable`.
+Removing that `Copyable` requirement for the nested generic context `S<T>.f` is useless, as there will never be a noncopyable value substituted for `S<T>`.
+The same logic applies to mututally-scoped contexts:
+
+```swift
+protocol P {
+  // error: cannot suppress constraint 'Self.Alice: ~Copyable' on generic parameter 'Self.Alice' defined in outer scope
+  associatedtype Bob where Alice: ~Copyable
+  associatedtype Alice where Bob: ~Copyable
+  // error: cannot suppress constraint 'Self.Bob: ~Copyable' on generic parameter 'Self.Bob' defined in outer scope
+}
+```
+
+## Source Compatibility
+
+Since `Copyable` is implicitly assumed for any context that does not explicitly
+specify `~Copyable`, this does not change the interpretation of existing code.
+
+## ABI compatibility
+
+Ordinarily, mangled symbols include a list of protocols to which the entity conforms.
+To preserve the ABI for existing code (which now implicitly conforms to `Copyable`),
+we are adjusting the mangling policy to encode the lack of a `Copyable` requirement
+rather than its presence.
+This ensures that existing ABI will end up with the same symbols as before.
+
+## Implications on adoption
+
+**ABI**: Adding `~Copyable` to an existing type, function, or generic parameter is generally an ABI-breaking change.
+
+Targeted mechanisms are being explored that can be used to preserve ABI compatibility when adding `~Copyable` to existing types,
+but such mechanisms will require extreme care to use correctly.
+In particular, this can allow clients to compile against new versions of a library and then run with an old version.
+Of course, if the old version attempts to copy a non-copyable value, this will break.
+
+**Source**: Adding `~Copyable` to a generic parameter on a function
+is generally not source-breaking for existing clients that provide
+implicitly-`Copyable` types for such parameters.
+
+Adding `~Copyable` to an existing type or protocol (generic or concrete) is typically
+source-breaking, as existing code may rely on the ability to copy values of this type.
+
+Adding `~Copyable` to an associated type is also generally source breaking.
+
+## Alternatives Considered
```

Comment:
I know there's been a lot of careful consideration around the defaults for protocols and their associated types. If there were any alternative directions that were considered in that area, e.g. not inferring protocols to refine `Copyable` by default, please specify them here and why those directions were not chosen.